Control/lib/services/Lock.js

```
@@ -0,0 +1,137 @@
+   */
+  lockDetector(req, res) {
+    try {
+      if (!('name' in req.body)) {
+        throw new Error('Unspecified lock entity');
+      }
+      const entity = req.body.name;
+      if (entity in this.lockedBy) {
+        throw new Error(`Lock ${entity} is already hold by ${this.lockedByName[entity]} (id ${this.lockedBy[entity]})`);
+      }   
+      this.lockedBy[entity] = req.session.personid;
+      this.lockedByName[entity] = req.session.name;
+      log.info(`Lock ${entity} taken by ${req.session.name}`);
+      this.broadcastLockState();
+      res.status(200).json({ok: true});
+    } catch (error) {
+      log.error(`Unable to lock by ${req.session.name}: ${error}`);
+      res.status(403).json({message: error.message});
+    }
+  }
+
+  /** 
+   * Method to try to release lock
+   * @param {Request} req
+   * @param {Response} res
+  */
+  forceUnlock(req, res) {
+    try {
+      if (!('name' in req.body)) {
+        throw new Error('Unspecified lock entity');
+      }
+      const entity = req.body.name;
+      if (!(entity in this.lockedBy)) {
+        throw new Error(`Lock ${entity} is already released`);
+      }   
+      if (!req.session.access.includes('admin')) {
+        throw new Error(`Insufficient permission`);
+      }   
+      delete this.lockedBy[entity];
+      delete this.lockedByName[entity];
+      log.info(`Lock ${entity} forced by ${req.session.name}`);
+      this.broadcastLockState();
+      res.status(200).json({ok: true});
+    } catch (error) {
+      log.error(`Unable to force lock by ${req.session.name}: ${error}`);
+      res.status(403).json({message: error.message});
+    }   
+  }
+
+  /** 
+   * Method to try to release lock
+   * @param {Request} req
+   * @param {Response} res
+   */
+  unlockDetector(req, res) {
+    try {
+      if (!('name' in req.body)) {
+        throw new Error('Unspecified lock entity');
+      }
+      const entity = req.body.name;
+      if (!(entity in this.lockedBy)) {
```

Do we want to throw an error if the lock is already released? In the end this is the expected behavior from the user; the action ends with the lock not being taken

---

Yes, indeed I agree, exceptions are dedicated for situations that are not expected, but in such case we shall remove almost all `throws` from Lock service.